File: net/JetBrains.SignatureVerifier/Crypt/SignedMessage.cs
using System;
using System.IO;
using JetBrains.Annotations;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Cms;
using Org.BouncyCastle.Cms;
using CmsSignedData = JetBrains.SignatureVerifier.BouncyCastle.Cms.CmsSignedData;

namespace JetBrains.SignatureVerifier.Crypt
{
  public class SignedMessage
  {
    private readonly CmsSignedData _cmsSignedData;

    public CmsSignedData SignedData => _cmsSignedData;


    public static SignedMessage CreateInstance(SignatureData signatureData)
    {
      if (signatureData.IsEmpty)
        throw new InvalidDataException(nameof(signatureData));

      if (signatureData.HasAttachedSignedData)
        return new SignedMessage(signatureData.SignedData, signatureData.CmsData);
      return new SignedMessage(signatureData.CmsData);
    }

    public SignedMessage([NotNull] byte[] pkcs7Data)
    {
      if (pkcs7Data == null) throw new ArgumentNullException(nameof(pkcs7Data));
      var _data = pkcs7Data ?? throw new ArgumentNullException(nameof(pkcs7Data));

      var asnStream = new Asn1InputStream(_data);
      var pkcs7 = ContentInfo.GetInstance(asnStream.ReadObject());
      _cmsSignedData = new CmsSignedData(pkcs7);
    }

    public SignedMessage([NotNull] byte[] signedData, [NotNull] byte[] pkcs7Data)
    {
      if (signedData == null) throw new ArgumentNullException(nameof(signedData));
      if (pkcs7Data == null) throw new ArgumentNullException(nameof(pkcs7Data));
      var signedContent = new CmsProcessableByteArray(signedData);

      try
      {
        using var asnStream = new Asn1InputStream(pkcs7Data);
        var pkcs7 = ContentInfo.GetInstance(asnStream.ReadObject());
        _cmsSignedData = new CmsSignedData(signedContent, pkcs7);
      }
      catch (IOException ex)
      {
        throw new Exception("Invalid signature format", ex);
      }
    }

    internal SignedMessage([NotNull] Asn1Object obj)
    {
      if (obj == null) throw new ArgumentNullException(nameof(obj));
      var pkcs7 = ContentInfo.GetInstance(obj);
      _cmsSignedData = new CmsSignedData(pkcs7);
    }
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/crypt/SignedMessage.kt
package com.jetbrains.signatureverifier.crypt

import com.jetbrains.signatureverifier.SignatureData
import org.bouncycastle.cms.CMSSignedData
import org.bouncycastle.asn1.ASN1InputStream
import org.bouncycastle.asn1.ASN1Object
import org.bouncycastle.asn1.cms.ContentInfo
import org.bouncycastle.cms.CMSProcessableByteArray
import org.jetbrains.annotations.NotNull
import java.io.IOException

class SignedMessage {
  companion object {
    fun CreateInstance(signatureData: SignatureData): SignedMessage {
      if (signatureData.IsEmpty)
        error("signatureData is empty")

      if (signatureData.HasAttachedSignedData)
        return SignedMessage(signatureData.SignedData!!, signatureData.CmsData!!)
      return SignedMessage(signatureData.CmsData!!)
    }
  }

  val SignedData: CMSSignedData

  private constructor (@NotNull pkcs7Data: ByteArray) {
    val asnStream = ASN1InputStream(pkcs7Data)
    val pkcs7 = ContentInfo.getInstance(asnStream.readObject())
    SignedData = CMSSignedData(pkcs7)
  }

  private constructor(@NotNull signedData: ByteArray, @NotNull pkcs7Data: ByteArray) {
    val signedContent = CMSProcessableByteArray(signedData)

    try {
      val asnStream = ASN1InputStream(pkcs7Data)
      val pkcs7 = ContentInfo.getInstance(asnStream.readObject())
      SignedData = CMSSignedData(signedContent, pkcs7)
    } catch (ex: IOException) {
      throw Exception("Invalid signature format", ex)
    }
  }

  internal constructor(@NotNull obj: ASN1Object) {
    val pkcs7 = ContentInfo.getInstance(obj)
    SignedData = CMSSignedData(pkcs7)
  }
}
